File: articles/stage-theories-of-development-promise-power-and-critique/r/model_stage_like_development.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "stage_theory_development_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_stage_theory_panel.py first.")

panel <- read.csv(data_path)

model <- lmer(
  development_score ~ time + current_support + school_support +
    resource_stability + chronic_stress + stage_pattern +
    threshold_on + logistic_transition +
    stage_pattern:threshold_on +
    stage_pattern:logistic_transition +
    stage_pattern:threshold_on:transition_readiness +
    (1 + time | context_id/child_id),
  data = panel
)

capture.output(
  summary(model),
  file = file.path(outputs_dir, "r_stage_model_summary.txt")
)

trajectory <- panel |>
  group_by(time, stage_pattern) |>
  summarize(
    average_score = mean(development_score),
    average_readiness = mean(transition_readiness),
    standard_error = sd(development_score) / sqrt(n()),
    .groups = "drop"
  ) |>
  mutate(
    lower = average_score - 1.96 * standard_error,
    upper = average_score + 1.96 * standard_error,
    pattern_label = ifelse(stage_pattern == 1, "Stage-like reorganization", "Mostly continuous growth")
  )

write.csv(trajectory, file.path(outputs_dir, "r_stage_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_stage_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_score, linetype = pattern_label)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper, group = pattern_label), alpha = 0.12) +
    labs(title = "Synthetic Continuous and Stage-Like Developmental Patterns", x = "Time", y = "Average development score", linetype = "Pattern") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

threshold_summary <- panel |>
  group_by(threshold_time, stage_pattern) |>
  summarize(
    average_score = mean(development_score),
    average_readiness = mean(transition_readiness),
    average_support = mean(current_support),
    average_stress = mean(chronic_stress),
    .groups = "drop"
  ) |>
  mutate(pattern_label = ifelse(stage_pattern == 1, "Stage-like reorganization", "Mostly continuous growth"))

write.csv(threshold_summary, file.path(outputs_dir, "r_threshold_summary.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_threshold_summary.png"),
  plot = ggplot(threshold_summary, aes(x = threshold_time, y = average_score, linetype = pattern_label)) +
    geom_line(linewidth = 1) +
    geom_point(size = 2) +
    labs(title = "Synthetic Threshold Timing and Developmental Outcome", x = "Threshold time", y = "Average development score", linetype = "Pattern") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R stage-like development outputs.")

File: articles/stage-theories-of-development-promise-power-and-critique/python/model_stage_like_development.py
#!/usr/bin/env python3
"""Model synthetic continuous and stage-like developmental trajectories."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "stage_theory_development_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_stage_theory_panel.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["child_id", "time"]).reset_index(drop=True)
    panel["lag_score"] = panel.groupby("child_id")["development_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    threshold_model = smf.ols(
        """
        development_score ~ lag_score + time + current_support +
        school_support + resource_stability + chronic_stress +
        threshold_on + logistic_transition + stage_pattern +
        transition_readiness + threshold_on:stage_pattern +
        logistic_transition:stage_pattern +
        threshold_on:stage_pattern:transition_readiness
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    profile_model = smf.ols(
        """
        development_score ~ time + transition_readiness + chronic_stress +
        C(stage_profile) + logistic_transition + stage_pattern
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_stage_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC STAGE-LIKE DEVELOPMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(threshold_model.summary().as_text())
        f.write("\n\nPROFILE MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(profile_model.summary().as_text())

    trajectory = panel.groupby(["time", "stage_pattern"], as_index=False).agg(
        average_score=("development_score", "mean"),
        average_readiness=("transition_readiness", "mean"),
        average_support=("current_support", "mean"),
        standard_error=("development_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["pattern_label"] = trajectory["stage_pattern"].map({
        0: "Mostly continuous growth",
        1: "Stage-like reorganization",
    })
    trajectory["lower"] = trajectory["average_score"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_score"] + 1.96 * trajectory["standard_error"]
    trajectory.to_csv(OUTPUTS_DIR / "python_stage_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    for group_name, group in trajectory.groupby("pattern_label"):
        plt.plot(group["time"], group["average_score"], marker="o", label=str(group_name))
    plt.xlabel("Time")
    plt.ylabel("Average development score")
    plt.title("Synthetic Continuous and Stage-Like Developmental Patterns")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_stage_trajectory.png", dpi=160)
    plt.close()

    threshold_summary = panel.groupby(["threshold_time", "stage_pattern"], as_index=False).agg(
        average_score=("development_score", "mean"),
        average_readiness=("transition_readiness", "mean"),
        average_support=("current_support", "mean"),
        average_stress=("chronic_stress", "mean"),
    )
    threshold_summary["pattern_label"] = threshold_summary["stage_pattern"].map({
        0: "Mostly continuous growth",
        1: "Stage-like reorganization",
    })
    threshold_summary.to_csv(OUTPUTS_DIR / "python_threshold_summary.csv", index=False)

    plt.figure(figsize=(8, 5))
    for group_name, group in threshold_summary.groupby("pattern_label"):
        plt.plot(group["threshold_time"], group["average_score"], marker="o", label=str(group_name))
    plt.xlabel("Threshold time")
    plt.ylabel("Average development score")
    plt.title("Synthetic Threshold Timing and Developmental Outcome")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_threshold_summary.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("stage_profile", as_index=False).agg(
        individuals=("child_id", "nunique"),
        average_score=("development_score", "mean"),
        average_readiness=("transition_readiness", "mean"),
        average_support=("current_support", "mean"),
        average_stress=("chronic_stress", "mean"),
        average_threshold=("threshold_time", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_stage_profiles.csv", index=False)

    print("Wrote Python stage-like development model outputs.")


if __name__ == "__main__":
    main()
